src/coreclr/vm/managedmdimport.cpp

```diff
@@ -613,49 +443,168 @@ MDImpl2(void, MetaDataImport::GetParentToken, mdToken tk, mdToken* ptk)
-    }
+    return hr;
 }
 FCIMPLEND
 
-MDImpl1(void, MetaDataImport::GetScopeProps, GUID* pmvid)
+MDImpl1(HRESULT, MetaDataImport::GetScopeProps, GUID* pmvid)
 {
     FCALL_CONTRACT;
 
-    HRESULT hr;
     LPCSTR szName;
-
     IMDInternalImport *_pScope = pScope;
-    hr = _pScope->GetScopeProps(&szName, pmvid);
-    if (FAILED(hr))
-    {
-        FCThrowVoid(kBadImageFormatException);
-    }
+    return _pScope->GetScopeProps(&szName, pmvid);
 }
 FCIMPLEND
 
 
-MDImpl2(void, MetaDataImport::GetMemberRefProps,
+MDImpl2(HRESULT, MetaDataImport::GetMemberRefProps,
     mdMemberRef mr,
     ConstArray* ppvSigBlob)
 {
     FCALL_CONTRACT;
 
-    HRESULT hr;
     IMDInternalImport *_pScope = pScope;
     LPCSTR szName_Ignore;
 
-    hr = _pScope->GetNameAndSigOfMemberRef(mr, (PCCOR_SIGNATURE*)&ppvSigBlob->m_array, (ULONG*)&ppvSigBlob->m_count, &szName_Ignore);
-    if (FAILED(hr))
-    {
-        FCThrowVoid(kBadImageFormatException);
-    }
+    return _pScope->GetNameAndSigOfMemberRef(mr, (PCCOR_SIGNATURE*)&ppvSigBlob->m_array, (ULONG*)&ppvSigBlob->m_count, &szName_Ignore);
 }
 FCIMPLEND
 
 #if defined(_MSC_VER) && defined(TARGET_X86)
-#pragma optimize("", on)			// restore command line optimization defaults
+#pragma optimize("", on)    // restore command line optimization defaults
 #endif
 
+struct ResultMemory final
+{
+    INT32 length;
+    INT32* alloc;
+
+    ~ResultMemory()
+    {
+        if (alloc != NULL)
+            ::free(alloc);
+    }
+
+    void AllocateManagedArray(QCall::ObjectHandleOnStack& longResult)
+    {
+        CONTRACTL
+        {
+            THROWS;
+            MODE_PREEMPTIVE;
+            PRECONDITION(alloc != NULL);
+        }
+        CONTRACTL_END;
+
+        {
+            GCX_COOP();
+            longResult.Set(AllocatePrimitiveArray(ELEMENT_TYPE_I4, length));
+            void* p = ((I4Array*)OBJECTREFToObject(longResult.Get()))->GetDirectPointerToNonObjectElements();
+            memcpyNoGCRefs(p, alloc, length * sizeof(INT32));
+        }
+    }
+};
+
+static void* EnsureResultSize(
+    INT32 resultLength,
+    INT32 shortResultLen,
+    INT32* shortResult,
+    ResultMemory& resultMemory)
+{
+    CONTRACT(void*)
+    {
+        THROWS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(shortResultLen > 0);
+        PRECONDITION(shortResult != NULL);
+        POSTCONDITION((RETVAL != NULL));
+    }
+    CONTRACT_END;
+
+    void* p;
+    INT32 resultInBytes;
+    if (resultLength <= shortResultLen)
+    {
+        resultInBytes = shortResultLen * sizeof(INT32);
+        p = shortResult;
+    }
+    else
+    {
+        _ASSERTE(resultLength > 0);
+        resultInBytes = resultLength * sizeof(INT32);
+
+        resultMemory.length = resultLength;
+        resultMemory.alloc = (INT32*)::malloc(resultInBytes);
+        if (resultMemory.alloc == NULL)
+            ThrowOutOfMemory();
+        p = resultMemory.alloc;
+    }
+    ZeroMemory(p, resultInBytes);
+    RETURN p;
+}
```

Comment:
```suggestion
static INT32* EnsureResultSize(
    INT32 resultLength,
    INT32 shortResultLen,
    INT32* shortResult,
    ResultMemory& resultMemory)
{
    CONTRACT(void*)
    {
        THROWS;
        MODE_PREEMPTIVE;
        PRECONDITION(shortResultLen > 0);
        PRECONDITION(shortResult != NULL);
        POSTCONDITION((RETVAL != NULL));
    }
    CONTRACT_END;

    void* p;
    if (resultLength <= shortResultLen)
    {
        p = shortResult;
    }
    else
    {
        _ASSERTE(resultLength > 0);

        resultMemory.length = resultLength;
        resultMemory.alloc = new INT32[resultLength];
        p = resultMemory.alloc;
    }
    ZeroMemory(p, (size_t)resultLength * sizeof(INT32));
    RETURN p;
}
```
I think this is trying to be too low-level for no good reason. `INT32*` and regular `new` should work just fine.

Reply:
I don't follow the argument here. Is this about the use of `malloc` and `free`? I don't consider that any more low level than manual `new` and `delete[]`. In this case using those is more complicated because the allocator isn't clear and not using `delete[]` is a common mistake. Another aspect of the suggestion is returning `void*` was intentional based on usage. I want to force an explicit cast in all use cases due to use of `ASSOCIATE_RECORD*` for `GetAllAssociates()` call.

I'm fine adopting the suggestion, however it doesn't seem all that less low level other than avoiding the explicit OOM call and implies a fair bit more nuance than what is presently very clear and necessary to be aware of.

Reply:
Consistency. We use new/delete everywhere else in vm (except for a few outliers).

My low-level comment was about computing byte sizes. `new` computes the size for you.

`resultLength * sizeof(INT32)` will get flagged by static analyzers for potential integer overflow.

Back in the day, we had a lot of places that computed bytes lengths like this and all of them get flagged as potential integer overflow warnings by static analysis tools. Our preferred strategy to fix them was to change to code to use `new`.

Reply:
> resultLength * sizeof(INT32) will get flagged by static analyzers for potential integer overflow.

I'm still not following the argument then. Changing this isn't going to get rid of that since we will do this for `ZeroMemory` and where we `memcpy` the memory as well. Is the concern about passing the computed size to an allocating function?

Reply:
The static analyzers have been looking specifically for result of unchecked multiplication being passed to an allocation method like malloc. They do not flag unchecked multiplications passed to ZeroMemory, etc.